Unintended breaking change with IsSameSequenceAs

From the release notes for 9.0.0:
> Breaking change : the signature of IsSameSequenceAs has [changed](https://github.com/FakeItEasy/FakeItEasy/commit/e815b77dfba2ef9c4e91fe35bc24f7910f334f87#diff-86b0fcd7c9ae751d623bf244dba6837b3393cb0c3c78a74ebf8ca316ddf209ecR73). It now has 2 generic type parameters, and accepts an optional equality comparer. This is a binary-level breaking change, and possibly a source-level breaking change, but we don't expect most users to be affected, since the generic type arguments are typically inferred by the compiler (https://github.com/FakeItEasy/FakeItEasy/issues/1913)

However, on .NET 9, it can break with: `An expression tree may not contain a call or invocation that uses optional arguments`

Sample repro:
```
using FakeItEasy;

ITest test = null!;
IEnumerable<string> strings = [];
A.CallTo(() => test.DoThings(A<IEnumerable<string>>.That.IsSameSequenceAs(strings)));

public interface ITest
{
    void DoThings(IEnumerable<string> things);
}
```

FYI: This compiles fine with .NET 10

OK, I think I got it.
Our BDD specs only cover the `params` overload. That wasn't necessarily our intention, but it's the overload that was resolved.
And our unit tests do test both overloads, but not within an expression.
So we never saw the issue...

And the reason it works in .NET 10 is that apparently, the "no optional args in expressions" limitation has been removed in C# 14. I don't remember seeing any announcement about this, but it seems to be the case. This code works fine in .NET 10 with the default language version, i.e. 14:
```csharp
    class Program
    {
        static void Main(string[] args)
        {
            Expression<Func<int>> expr = () => Foo();
        }

        static int Foo(string? s = null) => 42;
    }
```
But if you change `LangVersion` to 13, you get this:

> error CS0854: An expression tree may not contain a call or invocation that uses optional arguments

Anyway, the fix should be pretty easy. Just split the method in two overloads, with and without the comparer.

--- a/src/FakeItEasy/ArgumentConstraintManagerExtensions.cs
+++ b/src/FakeItEasy/ArgumentConstraintManagerExtensions.cs
@@ -213,6 +213,23 @@ public static class ArgumentConstraintManagerExtensions
     /// </summary>
     /// <param name="manager">The constraint manager to match the constraint.</param>
     /// <param name="values">The sequence to test against.</param>
+    /// <typeparam name="T">The type of argument to constrain.</typeparam>
+    /// <typeparam name="TElement">The type of the collection's elements.</typeparam>
+    /// <returns>A dummy argument value.</returns>
+    public static T IsSameSequenceAs<T, TElement>(
+        this IArgumentConstraintManager<T> manager,
+        IEnumerable<TElement> values)
+        where T : IEnumerable<TElement>
+    {
+        return manager.IsSameSequenceAs(values, null);
+    }
+
+    /// <summary>
+    /// Constrains the argument collection so that it must contain the same elements as the
+    /// specified collection, in the same order.
+    /// </summary>
+    /// <param name="manager">The constraint manager to match the constraint.</param>
+    /// <param name="values">The sequence to test against.</param>
     /// <param name="comparer">A comparer to test the collection elements for equality. If null, the default equality comparer for <c>TElement</c> will be used.</param>
     /// <typeparam name="T">The type of argument to constrain.</typeparam>
     /// <typeparam name="TElement">The type of the collection's elements.</typeparam>
@@ -220,7 +237,7 @@ public static class ArgumentConstraintManagerExtensions
     public static T IsSameSequenceAs<T, TElement>(
         this IArgumentConstraintManager<T> manager,
         IEnumerable<TElement> values,
-        IEqualityComparer<TElement>? comparer = null)
+        IEqualityComparer<TElement>? comparer)
         where T : IEnumerable<TElement>
     {
         Guard.AgainstNull(manager);
@@ -258,6 +275,23 @@ public static class ArgumentConstraintManagerExtensions
     /// </summary>
     /// <param name="manager">The constraint manager to match the constraint.</param>
     /// <param name="values">The sequence to test against.</param>
+    /// <typeparam name="T">The type of argument to constrain.</typeparam>
+    /// <typeparam name="TElement">The type of the collection's elements.</typeparam>
+    /// <returns>A dummy argument value.</returns>
+    public static T HasSameElementsAs<T, TElement>(
+        this IArgumentConstraintManager<T> manager,
+        IEnumerable<TElement> values)
+        where T : IEnumerable<TElement>
+    {
+        return manager.HasSameElementsAs(values, null);
+    }
+
+    /// <summary>
+    /// Constrains the argument collection so that it must contain the same elements as the
+    /// specified collection, in any order.
+    /// </summary>
+    /// <param name="manager">The constraint manager to match the constraint.</param>
+    /// <param name="values">The sequence to test against.</param>
     /// <param name="comparer">A comparer to test the collection elements for equality. If null, the default equality comparer for <c>TElement</c> will be used.</param>
     /// <typeparam name="T">The type of argument to constrain.</typeparam>
     /// <typeparam name="TElement">The type of the collection's elements.</typeparam>
@@ -265,7 +299,7 @@ public static class ArgumentConstraintManagerExtensions
     public static T HasSameElementsAs<T, TElement>(
         this IArgumentConstraintManager<T> manager,
         IEnumerable<TElement> values,
-        IEqualityComparer<TElement>? comparer = null)
+        IEqualityComparer<TElement>? comparer)
         where T : IEnumerable<TElement>
     {
         Guard.AgainstNull(manager);

--- a/tests/FakeItEasy.Tests.Approval/ApprovedApi/FakeItEasy/net10.0.verified.txt
+++ b/tests/FakeItEasy.Tests.Approval/ApprovedApi/FakeItEasy/net10.0.verified.txt
@@ -51,9 +51,11 @@ namespace FakeItEasy
             where T : System.Collections.IEnumerable { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value) { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value, System.StringComparison comparisonType) { }
+        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values)
+            where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, params TElement[] values)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
-        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer = null)
+        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static System.Threading.CancellationToken IsCanceled(this FakeItEasy.IArgumentConstraintManager<System.Threading.CancellationToken> manager) { }
         public static T IsEmpty<T>(this FakeItEasy.IArgumentConstraintManager<T> manager)
@@ -74,9 +76,11 @@ namespace FakeItEasy
             where T :  struct { }
         public static string IsNullOrEmpty(this FakeItEasy.IArgumentConstraintManager<string> manager) { }
         public static T IsSameAs<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, T value) { }
+        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values)
+            where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, params TElement[] values)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
-        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer = null)
+        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T Matches<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Linq.Expressions.Expression<System.Func<T, bool>> predicate) { }
         public static T Matches<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Func<T, bool> predicate, string description) { }

--- a/tests/FakeItEasy.Tests.Approval/ApprovedApi/FakeItEasy/net462.verified.txt
+++ b/tests/FakeItEasy.Tests.Approval/ApprovedApi/FakeItEasy/net462.verified.txt
@@ -51,9 +51,11 @@ namespace FakeItEasy
             where T : System.Collections.IEnumerable { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value) { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value, System.StringComparison comparisonType) { }
+        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values)
+            where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, params TElement[] values)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
-        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer = null)
+        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static System.Threading.CancellationToken IsCanceled(this FakeItEasy.IArgumentConstraintManager<System.Threading.CancellationToken> manager) { }
         public static T IsEmpty<T>(this FakeItEasy.IArgumentConstraintManager<T> manager)
@@ -74,9 +76,11 @@ namespace FakeItEasy
             where T :  struct { }
         public static string IsNullOrEmpty(this FakeItEasy.IArgumentConstraintManager<string> manager) { }
         public static T IsSameAs<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, T value) { }
+        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values)
+            where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, params TElement[] values)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
-        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer = null)
+        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T Matches<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Linq.Expressions.Expression<System.Func<T, bool>> predicate) { }
         public static T Matches<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Func<T, bool> predicate, string description) { }

--- a/tests/FakeItEasy.Tests.Approval/ApprovedApi/FakeItEasy/net8.0.verified.txt
+++ b/tests/FakeItEasy.Tests.Approval/ApprovedApi/FakeItEasy/net8.0.verified.txt
@@ -51,9 +51,11 @@ namespace FakeItEasy
             where T : System.Collections.IEnumerable { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value) { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value, System.StringComparison comparisonType) { }
+        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values)
+            where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, params TElement[] values)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
-        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer = null)
+        public static T HasSameElementsAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static System.Threading.CancellationToken IsCanceled(this FakeItEasy.IArgumentConstraintManager<System.Threading.CancellationToken> manager) { }
         public static T IsEmpty<T>(this FakeItEasy.IArgumentConstraintManager<T> manager)
@@ -74,9 +76,11 @@ namespace FakeItEasy
             where T :  struct { }
         public static string IsNullOrEmpty(this FakeItEasy.IArgumentConstraintManager<string> manager) { }
         public static T IsSameAs<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, T value) { }
+        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values)
+            where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, params TElement[] values)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
-        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer = null)
+        public static T IsSameSequenceAs<T, TElement>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Collections.Generic.IEnumerable<TElement> values, System.Collections.Generic.IEqualityComparer<TElement>? comparer)
             where T : System.Collections.Generic.IEnumerable<TElement> { }
         public static T Matches<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Linq.Expressions.Expression<System.Func<T, bool>> predicate) { }
         public static T Matches<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, System.Func<T, bool> predicate, string description) { }

